node/bindnode/fuzz_test.go

```diff
@@ -0,0 +1,238 @@
+		schemaDMT := bindnode.Unwrap(schemaNode).(*schemadmt.Schema)
+
+		// Log the input schema and node we're fuzzing with, to help debugging.
+		// We also use dag-json, as it's more human readable.
+		t.Logf("node in dag-cbor: %q", nodeDagCBOR)
+		t.Logf("schema in dag-json: %s", marshalDagJSON(t, schemaNode.Representation()))
+		{
+			nodeBuilder := basicnode.Prototype.Any.NewBuilder()
+			if err := dagcbor.Decode(nodeBuilder, bytes.NewReader(nodeDagCBOR)); err != nil {
+				// If some dag-cbor bytes don't decode into the Any prototype,
+				// then they're just not valid dag-cbor at all.
+				t.Skipf("invalid node dag-cbor: %v", err)
+			}
+			node := nodeBuilder.Build()
+			t.Logf("node in dag-json: %s", marshalDagJSON(t, node))
+		}
+
+		ts := new(schema.TypeSystem)
+		ts.Init()
+		// For the time being, we're not interested in panics from
+		// schemadmt.Compile or schema.TypeSystem. They are relatively prone to
+		// panics at the moment, and right now we're mainly interested in bugs
+		// in bindnode and dagcbor.
+		func() {
+			defer func() {
+				if r := recover(); r != nil {
+					t.Skipf("invalid schema: %v", r)
+				}
+			}()
+			schemadmt.Compile(ts, schemaDMT)
+		}()
+
+		schemaType := ts.TypeByName("Root")
+		if schemaType == nil {
+			t.Skipf("schema has no Root type")
+		}
+		var proto schema.TypedPrototype
+		func() {
+			defer func() {
+				if r := recover(); r != nil {
+					str := fmt.Sprint(r)
+					switch {
+					case strings.Contains(str, "bindnode: unexpected nil schema.Type"):
+					case strings.Contains(str, "is not a valid Go identifier"):
+					default:
+						panic(r)
+					}
+					t.Skipf("invalid schema: %v", r)
+				}
+			}()
+			proto = bindnode.Prototype(nil, schemaType)
+		}()
+
+		for _, repr := range []bool{false, true} {
+			t.Logf("decode and encode roundtrip with dag-cbor repr=%v", repr)
+			var nodeBuilder datamodel.NodeBuilder
+			if !repr {
+				nodeBuilder = proto.NewBuilder()
+			} else {
+				nodeBuilder = proto.Representation().NewBuilder()
+			}
+			if err := dagcbor.Decode(nodeBuilder, bytes.NewReader(nodeDagCBOR)); err != nil {
+				// The dag-cbor isn't valid for this node. Nothing else to do.
+				// We don't use t.Skip, because a dag-cbor might only be valid
+				// at the repr level, but not at the type level.
+				continue
+			}
+			node := nodeBuilder.Build()
+			if repr {
+				node = node.(schema.TypedNode).Representation()
+			}
+			// Unwrap returns a pointer, and %#v prints pointers as hex,
+			// so to get useful output, use reflect to dereference them.
+			t.Logf("decode successful: %#v", reflect.ValueOf(bindnode.Unwrap(node)).Elem().Interface())
+			reenc := marshalDagCBOR(t, node)
+			if !bytes.Equal(reenc, nodeDagCBOR) {
+				// TODO: reenable this once the dagcbor codec rejects non-canonical
+				// inputs like "00" and "a000", "0" and "a0" respectively.
+				// t.Errorf("node reencoded as %q rather than %q", reenc, nodeDagCBOR)
```

Comment:
I'm not sure if this matters for graphsync. The dagcbor codec will, as of today, happily accept non-canonical inputs like `a000`; note that it's the string `"0"`, but with two trailing bytes `"00"`. If it were a three-byte string, you'd have `c000`:

https://cbor.me/?diag=%220%22
https://cbor.me/?diag=%22000%22